### PR TITLE
Particle system

### PR DESCRIPTION
**Pull request by SeanAbnerNash**

### CODE REVIEW

File: ARGO/src/Entity.cpp [hunk 1]
```diff
@@ -222,16 +244,43 @@ void Entity::removeCompType(ComponentType t_type)
...
+				m_components.at(COMPONENT_ID::PRIMITIVE_ID) = nullptr;
+				return;
+			}
+			break;
+		}
+
 		case ComponentType::Force:
 		{
-			if (m_components.at(COMPONENT_ID::FORCE_ID))
+			if (!m_components.at(COMPONENT_ID::FORCE_ID))
```
Comment (CathalRedmond) [inline ARGO/src/Entity.cpp]: why did you had a '!' here?

File: ARGO/src/Entity.cpp [hunk 2]
```diff
@@ -88,6 +88,26 @@ void Entity::addComponent(Component* t_c)
 			}
 			break;
 		}
+
+		case ComponentType::Primitive:
```
Comment (CathalRedmond) [inline ARGO/src/Entity.cpp:92]: you didn't add primitive and particle component to the get component function

File: ARGO/src/Game.cpp [hunk 3]
```diff
@@ -49,9 +49,21 @@ Game::Game() :
...
 
 		//add components to player
+
+		
+		//int i = 0;  //Uncomment to test Particle Emmitter
 		for (auto& player : m_players)
 		{
 			createPlayer(player);
+
+			////Uncomment to test Particle Emmitter - Start
```
Comment (CathalRedmond) [inline ARGO/src/Game.cpp]: If you need to test it why not just add particles to player instead. Or make this a function.

File: ARGO/src/Game.cpp [hunk 4]
```diff
@@ -345,6 +358,8 @@ void Game::createPlayer(Entity& t_player)
 	t_player.addComponent(new ForceComponent());
 	t_player.addComponent(new ColliderCircleComponent(Utilities::PLAYER_RADIUS));
 	t_player.addComponent(new ColourComponent(glm::linearRand(0, 255), glm::linearRand(0, 255), glm::linearRand(0, 255), 255));
+	t_player.addComponent(new ParticleEmitterComponent(static_cast<TransformComponent*>(t_player.getComponent(ComponentType::Transform))->getPos(),true,90,30,1,100,20));
```
Comment (CathalRedmond) [inline ARGO/src/Game.cpp]: Magic numbers here at the end of particle emitter. Have no idea what they are just by looking at it

File: ARGO/src/Particle.cpp [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+///@Login C00217019
+
+Particle::Particle()
+{
+
+}
+
+Particle::~Particle()
+{
+
```
Comment (CathalRedmond) [inline ARGO/src/Particle.cpp:14]: Extra empty lines for no reason

File: ARGO/src/ParticleEmitterComponent.cpp [hunk 6]
```diff
@@ -0,0 +1,203 @@
+#include "stdafx.h"
+#include "ParticleEmitterComponent.h"
+
+ParticleEmitterComponent::ParticleEmitterComponent() :
+	Component(ComponentType::ParticleEmitter)
+{
+	for (int i = 0; i < m_maxParticles; i++)
+	{
+		Particle x;
```
Comment (CathalRedmond) [inline ARGO/src/ParticleEmitterComponent.cpp]: Ambiguous name 'x'.

Comment (CathalRedmond) [reply ARGO/src/ParticleEmitterComponent.cpp]: Same in overloaded constructor

File: ARGO/src/ParticleEmitterComponent.cpp [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+		m_particles.push_back(x);
+	}
+}
+
+///This Function Places Particles. Passing in a glm::vec2 Position will have it place the currently queued particle if possible. It will skip over it if it isn't.
+void ParticleEmitterComponent::setParticle(glm::vec2 t_pos)
+{
+	if (m_particles[m_placeParticle].getAlive() == false)//If the current particle to be tracked is dead then it places it then increments the tracker
```
Comment (CathalRedmond) [inline ARGO/src/ParticleEmitterComponent.cpp:40]: 'if(!m_particles[m_placeParticle].getAlive())' can be used, don't need '== false'

File: ARGO/include/ParticleEmitterComponent.h [hunk 8]
```diff
@@ -0,0 +1,61 @@
+#pragma once
```
Comment (CathalRedmond) [inline ARGO/include/ParticleEmitterComponent.h:1]: For this file in general:
1) Will you move the comments explaining what the functions do to the cpp. The way it is know in the header is cluttered. Also its better to have the explanation where the definition is so someone doesn't have to go back and forth to understand the function.
2) If it is just a simple getter/setter you don't need a comment. For example, setEmitting only sets the emitting variable so no comment is needed

File: ARGO/src/ParticleEmitterComponent.cpp [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+}
+//Sets the position of the Emitter. The point from which the particles emerge from
+void ParticleEmitterComponent::setPosition(glm::vec2 t_pos)
+{
+	m_position = t_pos;
+}
+
+bool ParticleEmitterComponent::isParticleDead(int t_index)
+{
+	if (m_particles[t_index].getTimer() > m_timeToKillParticle)
```
Comment (CathalRedmond) [inline ARGO/src/ParticleEmitterComponent.cpp]: you can just use 'return m_particles[t_index].getTimer() > m_timeToKillParticle', it does the same

File: ARGO/src/ParticleSystem.cpp [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+ParticleSystem::~ParticleSystem()
+{
+}
+
+void ParticleSystem::update(Entity& t_entity)
+{
+	if (t_entity.getAllComps().at(COMPONENT_ID::PARTICLE_ID) && t_entity.getAllComps().at(COMPONENT_ID::PRIMITIVE_ID) && t_entity.getAllComps().at(COMPONENT_ID::TRANSFORM_ID))
+	{
+		ParticleEmitterComponent* partComp = static_cast<ParticleEmitterComponent*>(t_entity.getAllComps().at(COMPONENT_ID::PARTICLE_ID));
```
Comment (CathalRedmond) [inline ARGO/src/ParticleSystem.cpp]: ambiguous name. kinda makes it seems like its part of a component

File: ARGO/src/ParticleSystem.cpp [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+		PrimitiveComponent* primtiveComp = static_cast<PrimitiveComponent*>(t_entity.getAllComps().at(COMPONENT_ID::PRIMITIVE_ID));
+		partComp->setPosition(static_cast<TransformComponent*>(t_entity.getAllComps().at(COMPONENT_ID::TRANSFORM_ID))->getPos());//Sets the Emitter to the entity's tranform position. This makes it move with mobile entities.
+		for (int i = 0; i < partComp->getMaxParticles(); i++)
+		{
+			//If the Emitter is emitting and the currently queded particle is not already alive
+			if (partComp->getEmitting() && !partComp->getParticleAlive(partComp->getNextParticle()))
+			{
+				//And it is time to place a new particle
+				if (partComp->getPlaceParticleTimer() > 1.0f)
```
Comment (CathalRedmond) [inline ARGO/src/ParticleSystem.cpp]: why not add this to previous if check

File: ARGO/src/ParticleSystem.cpp [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+					partComp->setParticleMovement(partComp->getNextParticle(), randomDirectionVectorInRange(partComp->getAngle(), partComp->getAngleOffset()));
+					partComp->setParticle(partComp->getEmitterPosition() + glm::vec2(randomOffset(partComp->getOffset()), partComp->getOffset()));
+					partComp->setPlaceParticleTimer(partComp->getPlaceParticleTimer() - 1);
+				}
+			}
+			//If the Current particle is alive.
+			if (partComp->getParticleAlive(i))
+			{
+				//Checks to see if the Particle should be killed
+				if (partComp->isParticleDead(i))
```
Comment (CathalRedmond) [inline ARGO/src/ParticleSystem.cpp]: Name of this function doesn't make sense. If isParticleDead returns true you would expect no need to have to kill it.

Comment (CathalRedmond) [reply ARGO/src/ParticleSystem.cpp]: Also if getParticleAlive returns true you shouldn't expect to be dead either

File: ARGO/src/ParticleSystem.cpp [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+			}
+		}
+		//Increment the placeTimer. This value is set by the particlesPerSecond variable. When the place timer is more than 1 a particle is placed.
+		partComp->setPlaceParticleTimer(partComp->getPlaceParticleTimer() + partComp->getParticlesPerSecond());
+	}
+}
+
+glm::vec2 ParticleSystem::randomDirectionVectorInRange(float t_angle, float t_angleOffset)
+{
+	std::random_device rd; // obtain a random number from hardware
```
Comment (CathalRedmond) [inline ARGO/src/ParticleSystem.cpp]: ambiguous name 'rd'

File: ARGO/src/ParticleSystem.cpp [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+	std::mt19937 eng(rd()); // seed the generator
```
Comment (CathalRedmond) [inline ARGO/src/ParticleSystem.cpp]: ambiguous name 'eng'

File: ARGO/src/ParticleSystem.cpp [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+	std::uniform_int_distribution<> distr(-t_angleOffset, t_angleOffset); // define the range
```
Comment (CathalRedmond) [inline ARGO/src/ParticleSystem.cpp]: ambiguous name 'distr'

File: ARGO/include/ParticleEmitterComponent.h (hunk 8 above)
Comment (CathalRedmond) [reply ARGO/include/ParticleEmitterComponent.h:1]: Same for ParticleSystem.h

**Review by CathalRedmond — CHANGES_REQUESTED** [review]

File: ARGO/src/Game.cpp (hunk 3 above)
Comment (SeanAbnerNash) [reply ARGO/src/Game.cpp]: I did. This is just the easiest place to set place to set the particles for players in specific ways.

File: ARGO/src/Game.cpp (hunk 4 above)
Comment (SeanAbnerNash) [reply ARGO/src/Game.cpp]: Stand in values in a constructor. Typing into the function will have pop up say what the expected inputs are. There aren't really any ways to label the values without making values.

File: ARGO/src/ParticleSystem.cpp (hunk 11 above)
Comment (SeanAbnerNash) [reply ARGO/src/ParticleSystem.cpp]: Readability.

File: ARGO/src/ParticleSystem.cpp (hunk 12 above)
Comment (SeanAbnerNash) [reply ARGO/src/ParticleSystem.cpp]: Redid name.

File: ARGO/src/ParticleSystem.cpp (hunk 13 above)
Comment (SeanAbnerNash) [reply ARGO/src/ParticleSystem.cpp]: Adjusted.

File: ARGO/src/Entity.cpp (hunk 1 above)
Comment (SeanAbnerNash) [reply ARGO/src/Entity.cpp]: Relic from a merge I missed. Fixed.

File: ARGO/src/ParticleEmitterComponent.cpp (hunk 7 above)
Comment (SeanAbnerNash) [reply ARGO/src/ParticleEmitterComponent.cpp:40]: Preference

File: ARGO/include/ParticleEmitterComponent.h (hunk 8 above)
Comment (SeanAbnerNash) [reply ARGO/include/ParticleEmitterComponent.h:1]: Both hpp. and cpp are commented so looking at either will explain the functions.

File: ARGO/src/ParticleEmitterComponent.cpp (hunk 9 above)
Comment (SeanAbnerNash) [reply ARGO/src/ParticleEmitterComponent.cpp]: Adjusted.

File: ARGO/src/ParticleSystem.cpp (hunk 10 above)
Comment (SeanAbnerNash) [reply ARGO/src/ParticleSystem.cpp]: Adjusted.

File: ARGO/src/Game.cpp (hunk 3 above)
Comment (SeanAbnerNash) [reply ARGO/src/Game.cpp]: Removed

File: ARGO/src/Game.cpp (hunk 4 above)
Comment (SeanAbnerNash) [reply ARGO/src/Game.cpp]: Sample Values created in Utilities

File: ARGO/src/ParticleEmitterComponent.cpp (hunk 6 above)
Comment (SeanAbnerNash) [reply ARGO/src/ParticleEmitterComponent.cpp]: Renamed

File: ARGO/include/ParticleEmitterComponent.h (hunk 8 above)
Comment (SeanAbnerNash) [reply ARGO/include/ParticleEmitterComponent.h:1]: Adjusted

File: ARGO/src/ParticleEmitterComponent.cpp [hunk 16]
```diff
@@ -0,0 +1,223 @@
...
+	m_maxParticles(t_maxParticles),
+	m_timeToKillParticle(t_timeToKillParticle),
+	m_emitting(t_emitting),
+	m_particlesPerSecond(t_particlesPerSecond / 60.0f),
+	m_rotating(t_rotating),
+	m_rotateAnglePerFrame(t_rotateAnglePerFrame)
+{
+	for (int i = 0; i < m_maxParticles; i++)
+	{
+		Particle particleX;
```
Comment (CathalRedmond) [inline ARGO/src/ParticleEmitterComponent.cpp]: you dont need a x here, just call it particle

Comment (CathalRedmond) [reply ARGO/src/ParticleEmitterComponent.cpp]: same for other constructor

**Review by CathalRedmond — CHANGES_REQUESTED** [review]

File: ARGO/src/ParticleEmitterComponent.cpp [hunk 17]
```diff
@@ -0,0 +1,223 @@
+#include "stdafx.h"
+#include "ParticleEmitterComponent.h"
+
+ParticleEmitterComponent::ParticleEmitterComponent() :
+	Component(ComponentType::ParticleEmitter)
+{
+	for (int i = 0; i < m_maxParticles; i++)
+	{
+		Particle particula;
```
Comment (CathalRedmond) [inline ARGO/src/ParticleEmitterComponent.cpp]: 'particle' not particula

Comment (CathalRedmond) [reply ARGO/src/ParticleEmitterComponent.cpp]: same for other constructor

**Review by CathalRedmond — CHANGES_REQUESTED** [review]

**Pull request merged**